packages/profile-sync-controller/src/AuthenticationController.ts

```diff
@@ -0,0 +1,338 @@
+import type {
+  RestrictedControllerMessenger,
+  StateMetadata,
+} from '@metamask/base-controller';
+import { BaseController } from '@metamask/base-controller';
+import type { InternalAccount } from '@metamask/keyring-api';
+import type { HandleSnapRequest } from '@metamask/snaps-controllers';
+
+import {
+  createSnapPublicKeyRequest,
+  createSnapSignMessageRequest,
+} from './AuthSnapRequests';
+import {
+  createLoginRawMessage,
+  getAccessToken,
+  getNonce,
+  login,
+} from './services/authentication-controller';
+import type { UserStorageControllerDisableProfileSyncing } from './UserStorageController';
+
+const THIRTY_MIN_MS = 1000 * 60 * 30;
+
+const controllerName = 'AuthenticationController';
+
+type SessionProfile = {
+  identifierId: string;
+  profileId: string;
+};
+
+type SessionData = {
+  /** profile - anonymous profile data for the given logged in user */
+  profile: SessionProfile;
+  /** accessToken - used to make requests authorized endpoints */
+  accessToken: string;
+  /** expiresIn - string date to determine if new access token is required  */
+  expiresIn: string;
+};
+
+type MetaMetricsAuth = {
+  getMetaMetricsId: () => string;
+};
+
+export type AuthenticationControllerState = {
+  /**
+   * Global isSignedIn state.
+   * Can be used to determine if "Profile Syncing" is enabled.
+   */
+  isSignedIn: boolean;
+  sessionData?: SessionData;
+  internalAccounts?: {
+    accounts: Record<string, InternalAccount>;
+    selectedAccount: string;
+  };
+};
+const defaultState: AuthenticationControllerState = { isSignedIn: false };
+const metadata: StateMetadata<AuthenticationControllerState> = {
+  isSignedIn: {
+    persist: true,
+    anonymous: true,
+  },
+  sessionData: {
+    persist: true,
+    anonymous: false,
+  },
+};
+
+// Messenger Actions
+type CreateActionsObj<T extends keyof AuthenticationController> = {
+  [K in T]: {
+    type: `${typeof controllerName}:${K}`;
+    handler: AuthenticationController[K];
+  };
+};
+type ActionsObj = CreateActionsObj<
+  | 'performSignIn'
+  | 'performSignOut'
+  | 'getBearerToken'
+  | 'getSessionProfile'
+  | 'isSignedIn'
+>;
+export type Actions = ActionsObj[keyof ActionsObj];
+export type AuthenticationControllerPerformSignIn = ActionsObj['performSignIn'];
+export type AuthenticationControllerPerformSignOut =
+  ActionsObj['performSignOut'];
+export type AuthenticationControllerGetBearerToken =
+  ActionsObj['getBearerToken'];
+export type AuthenticationControllerGetSessionProfile =
+  ActionsObj['getSessionProfile'];
+export type AuthenticationControllerIsSignedIn = ActionsObj['isSignedIn'];
```

Comment:
A few notes:

- What are your thoughts on being explicit with the actions? I realize that it seems like a lot of code, but much of it is copy-paste, and it would be good to use a common pattern across all our controllers so that when we're looking at this we don't have to learn how to read it
- Action types should end with "Action"
- The actions union type should start with the controller name

What about:

``` suggestion
export type AuthenticationControllerPerformSignInAction = {
  type: `${typeof controllerName}:performSignInAction`;
  handler: AuthenticationController['performSignInAction'];
};

export type AuthenticationControllerPerformSignOutAction = {
  type: `${typeof controllerName}:performSignOutAction`;
  handler: AuthenticationController['performSignOutAction'];
};

export type AuthenticationControllerGetBearerTokenAction = {
  type: `${typeof controllerName}:getBearerToken`;
  handler: AuthenticationController['getBearerToken'];
};

export type AuthenticationControllerGetSessionProfileAction = {
  type: `${typeof controllerName}:getSessionProfile`;
  handler: AuthenticationController['getSessionProfile'];
};

export type AuthenticationControllerIsSignedInAction = {
  type: `${typeof controllerName}:isSignedIn`;
  handler: AuthenticationController['isSignedIn'];
};

export type AuthenticationControllerActions =
  | AuthenticationControllerPerformSignInAction
  | AuthenticationControllerPerformSignOutAction
  | AuthenticationControllerGetBearerTokenAction
  | AuthenticationControllerGetSessionProfileAction
  | AuthenticationControllerIsSignedInAction;